File: clients/javascript/src/selenium-session-client.js
import fetch from 'node-fetch';

const DEFAULT_SESSION_SERVER = 'http://localhost:3000';
let sessionServer = null;

/**
 * get cookies from session server according to given tags
 * @param {*} sessionServer - session server url
 * @param {*} tags - tags by which to get cookies
 * @returns 
 */
const getSessionCookies = async (sessionServer, tags) => {
  let tagsTxt = tags ? `?tags=${tags.join(',')}` : '';
  let cookies = await fetch(`${sessionServer}/api/session${tagsTxt}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });
  return await cookies.json();
};

/**
 * set cookies to session server according to given cookies
 * @param {*} driver - selenium webdriver
 * @param {*} cookies - cookies to set
 */
const setSessionCookies = async (driver, cookies) => {
  await driver.sendAndGetDevToolsCommand('Network.clearBrowserCookies');
  await driver
    .sendAndGetDevToolsCommand('Network.setCookies', { cookies: cookies })
    .then(() => {
      console.log('cookies has been set');
    });
}

/**
 * Set session server URL
 * @param {*} server 
 */
const setSessionServer = (server) => {
  sessionServer = server;
}

/**
 * Get session server URL
 * @returns session server url
 */
const getSessionServer = () => {
  return sessionServer || process.env.SESSION_SERVER || DEFAULT_SESSION_SERVER;
}

/**
 * Init webdriver session with cookies from session server according to given tags
 * @param {*} driver - selenium webdriver to set cookies to
 * @param {*} tags - tags by which to get cookies
 */
const initSession = async (driver, tags) => {
  let cookies = await getSessionCookies(getSessionServer(), tags);
  await setSessionCookies(driver, cookies);
}

export { getSessionCookies, setSessionCookies, setSessionServer, getSessionServer, initSession };

File: clients/csharp/src/SeleniumSessionClient/SeleniumSessionClient.cs
using OpenQA.Selenium.Chrome;
using Newtonsoft.Json;
using System.Threading.Tasks;
using System.Net.Http;
using System.Collections.Generic;

namespace SeleniumSessionClient;
public class SeleniumSessionClient
{
    private static string DEFAULT_SESSION_SERVER = "http://localhost:3000";
    private static string? sessionServer = null;

    /// <summary>
    /// Set the session server address.
    /// </summary>
    /// <param name="server">The session server address.</param>
    public static void setSessionServer(string server)
    {
        sessionServer = server;
    }

    /// <summary>
    /// Get the session server address.
    /// </summary>
    /// <returns>The session server address.</returns>
    public static string getSessionServer()
    {
        if (sessionServer != null)
        {
            return sessionServer;
        }
        sessionServer = System.Environment.GetEnvironmentVariable("SESSION_SERVER");
        if (sessionServer != null)
        {
            return sessionServer;
        }
        return DEFAULT_SESSION_SERVER;
    }

    /// <summary>
    /// Get cookies from session server according to given tags
    /// </summary>
    /// <param name="sessionServer">url of session server</param>
    /// <param name="tags">array of tags according to which to get the cookies</param>
    /// <returns>JSON with cookies</returns>
    public static async Task<object> GetSessionCookies(string sessionServer, string[] tags)
    {
        string tagsParam = tags == null ? "" : $"?tags={string.Join(" ", tags)}";
        string url = $"{sessionServer}/api/session{tagsParam}";
        HttpClient client = new HttpClient();
        HttpResponseMessage response = await client.GetAsync(url);
        response.EnsureSuccessStatusCode();
        string responseBody = await response.Content.ReadAsStringAsync();
        return JsonConvert.DeserializeObject<object>(responseBody);
    }

    /// <summary>
    /// Set driver cookies according to given cookies
    /// </summary>
    /// <param name="driver">driver for which to set cookies</param>
    /// <param name="cookies">cookies to set</param>
    public static void SetSessionCookies(ChromeDriver driver, object cookies)
    {
        Dictionary<string, object> cdpCookies = new Dictionary<string, object>();
        cdpCookies.Add("cookies", cookies);
        driver.ExecuteCdpCommand("Network.setCookies", cdpCookies);
    }

    /// <summary>
    /// Initialize session of given driver with cookies from session server according to given tags
    /// </summary>
    /// <param name="driver">driver for which to set cookies</param>
    /// <param name="tags">tags by which to get cookies to set from session server</param>
    public static async Task initSession(ChromeDriver driver, string[] tags)
    {
        var cookies = await GetSessionCookies(getSessionServer(), tags);
        SetSessionCookies(driver, cookies);
    }
}